File: data structures and algorithms/cpp/03_008.cpp
#include <iostream>
#include <vector>
#include <string>

std::vector<std::string> get_permutations(
        const std::string& str) {
    // Base case
    if (str.length() <= 1) {
        return {str};
    }
    
    // Recursive case
    std::vector<std::string> perms;
    for (size_t i = 0; i < str.length(); i++) {
        char ch = str[i];
        std::string remaining = str.substr(0, i) 
                              + str.substr(i + 1);
        
        for (const std::string& p : get_permutations(remaining)) {
            perms.push_back(ch + p);
        }
    }
    
    return perms;
}

// Example usage
int main() {
    std::vector<std::string> result = get_permutations("abc");
    for (const std::string& perm : result) {
        std::cout << perm << " ";
    }
    std::cout << std::endl;  // Outputs: abc acb bac bca cab cba
    
    return 0;
}

File: data structures and algorithms/cs/03_008.cs
using System;
using System.Collections.Generic;

public class StringPermutations {
   public static List<string> GetPermutations(string str) {
      // Base case
      List<string> result = new List<string>();
      if (str.Length <= 1) {
         result.Add(str);
         return result;
      }
      
      // Recursive case
      for (int i = 0; i < str.Length; i++) {
         char currentChar = str[i];
         string remaining = str.Substring(0, i) + 
            str.Substring(i + 1);
         
         foreach (string perm in GetPermutations(remaining)) {
            result.Add(currentChar + perm);
         }
      }
      
      return result;
   }
   
   public static void Main(string[] args) {
      List<string> result = GetPermutations("abc");
      Console.WriteLine(string.Join(", ", result));
      // Output: abc, acb, bac, bca, cab, cba
   }
}